Orange/widgets/unsupervised/owlouvainclustering.py

```diff
@@ -0,0 +1,361 @@
+    else:
+        class Outputs:
+            annotated_data = Output(ANNOTATED_DATA_SIGNAL_NAME, Table, default=True)
+
+    apply_pca = ContextSetting(True)
+    pca_components = ContextSetting(_DEFAULT_PCA_COMPONENTS)
+    metric_idx = ContextSetting(0)
+    k_neighbors = ContextSetting(_DEFAULT_K_NEIGHBORS)
+    resolution = ContextSetting(1.)
+    auto_commit = Setting(True)
+
+    class Error(widget.OWWidget.Error):
+        empty_dataset = Msg('No features in data')
+        general_error = Msg('Error occured during clustering\n{}')
+
+    class State(Enum):
+        Pending, Running = range(2)
+
+    def __init__(self):
+        super().__init__()
+
+        self.data = None  # type: Optional[Table]
+        self.preprocessed_data = None  # type: Optional[Table]
+        self.graph = None  # type: Optional[nx.Graph]
+        self.partition = None  # type: Optional[np.array]
+
+        self.__executor = ThreadExecutor(parent=self)
+        self.__future = None  # type: Optional[Future]
+        self.__state = self.State.Pending
+
+        pca_box = gui.vBox(self.controlArea, 'PCA Preprocessing')
+        self.apply_pca_cbx = gui.checkBox(
+            pca_box, self, 'apply_pca', label='Apply PCA preprocessing',
+            callback=self._update_apply_pca,
+        )  # type: QCheckBox
+        self.pca_components_slider = gui.hSlider(
+            pca_box, self, 'pca_components', label='Components: ', minValue=2,
+            maxValue=_MAX_PCA_COMPONENTS,
+            callback=self._update_pca_components,
+        )  # type: QSlider
+
+        graph_box = gui.vBox(self.controlArea, 'Graph parameters')
+        self.metric_combo = gui.comboBox(
+            graph_box, self, 'metric_idx', label='Distance metric',
+            items=[m[0] for m in METRICS], callback=self._update_metric,
+            orientation=Qt.Horizontal,
+        )  # type: gui.OrangeComboBox
+        self.k_neighbors_spin = gui.spin(
+            graph_box, self, 'k_neighbors', minv=1, maxv=_MAX_K_NEIGBOURS,
+            label='k neighbors', controlWidth=80, alignment=Qt.AlignRight,
+            callback=self._update_k_neighbors,
+        )  # type: gui.SpinBoxWFocusOut
+        self.cls_epsilon_spin = gui.spin(
+            graph_box, self, 'resolution', 0, 5., 1e-2, spinType=float,
+            label='Resolution', controlWidth=80, alignment=Qt.AlignRight,
+            callback=self._update_resolution,
+        )  # type: gui.SpinBoxWFocusOut
+
+        self.apply_button = gui.auto_commit(
+            self.controlArea, self, 'auto_commit', 'Apply', box=None,
+            commit=self.commit,
+        )  # type: QWidget
+
+    def _update_apply_pca(self):
+        self._invalidate_graph()
+        self.commit()
+
+    def _update_pca_components(self):
+        self._invalidate_pca_projection()
+        self.commit()
+
+    def _update_metric(self):
+        self._invalidate_graph()
+        self.commit()
+
+    def _update_k_neighbors(self):
```

Comment:
As long as you do not expect these methods will be expanded in the future with code specific to changing one of the settings I don't see a reason to have 3 methods with the same body but different names.
I think having just one _update_graph and using it as the callback for all 3 controls would be better.